### PR TITLE
electricitymap and carbon intensity api clients

### PR DESCRIPTION
**Pull request by ofpiyush**
Adds basic energymap and carbonintensity API clients with a common interface with tests.

### CODE REVIEW

File: energymap/client.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+	return a, nil
+}
+
+type ApiClient struct {
+	client             *http.Client
+	apiURL             string
+	token              string
+	moderate_threshold float64
+	high_threshold     float64
```
Comment (rossf7) [inline energymap/client.go:50]:
```suggestion
	moderateThreshold float64
	highThreshold     float64
```

Could we use the same casing as the other fields?

File: energymap/errors.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+)
+
+var (
+	ErrNoRegionProvided           = errors.New("no region was provided")
```
Comment (rossf7) [inline energymap/errors.go:11]:
```suggestion
	ErrNoRegionProvided          error = errors.New("no region was provided")
```

File: energymap/api_options.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import "net/http"
+
+func WithHTTPClient(h *http.Client) ApiOption {
+	return func(a *ApiClient) error {
+		a.client = h
+		return nil
+	}
+}
+
+func WithThresholds(moderate, high float64) ApiOption {
```
Comment (rossf7) [inline energymap/api_options.go:12]: This part isn't clear to me. What are these thresholds used for? Also what about adding a comment explaining their purpose?

File: examples/energymap/main.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+	"context"
+	"log"
+	"os"
+
+	gridintensity "github.com/thegreenwebfoundation/grid-intensity-go"
+	"github.com/thegreenwebfoundation/grid-intensity-go/energymap"
+)
+
+func main() {
+	token := os.Getenv("ENERGY_MAP_API_TOKEN")
```
Comment (rossf7) [inline examples/energymap/main.go:13]: How can you get an API token for the Energy Map API?

**Review by rossf7 — COMMENTED** [review]
@ofpiyush Thanks this is great. I like the clean structure and that no external dependencies are needed.

I just have a couple of questions and some suggestions on minor things.

**Comment by ofpiyush** [comment]
Thank you for reviewing this!!

I'll check this after work, in about ~4 hours from now.

File: carbonintensity/schema.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package carbonintensity
+
+type CarbonIntensityResponse struct {
+	Data []IntensityData `json:"data"`
+}
+
+type IntensityData struct {
+	//From      time.Time `json:"from"`
+	//To        time.Time `json:"to"`
```
[the remark this answers is not in the document]
Comment (ofpiyush) [reply carbonintensity/schema.go:9]: No, they're not. I'll remove them.

File: energymap/client.go (hunk 1 above)
Comment (ofpiyush) [reply energymap/client.go:50]: We don't actually need these as well. This is dangling code from before.

File: energymap/api_options.go (hunk 3 above)
Comment (ofpiyush) [reply energymap/api_options.go:12]: Short answer: dangling code, I'll remove it.

Long answer for all of these: on main thread

File: examples/energymap/main.go (hunk 4 above)
Comment (ofpiyush) [reply examples/energymap/main.go:13]: answer on main thread

**Comment by ofpiyush** [comment]
## EnergyMap

It's a typo, it's actually supposed to be electrictymap, I'll rename everything now. 🙈 

For the API key,

You need to buy it from them at: https://api.electricitymap.org/#pricing

As an OSS alternative, I made a very hacked together project based on their OSS parsers for testing. I'll eventually clean it out a little and make a release from it so that people can self-host and get slightly lower quality data than the actual paid API.


In it's current state, it works in the happy path.

https://github.com/ofpiyush/carbonintensity-api/pull/1

## All dangling code.

We initially started with trying to mimic what https://github.com/thegreenwebfoundation/grid-intensity does.

It gives you back a value like "low", "high" for the carbon index.

That's because the carbonintensity.co.uk endpoint actually returns one. So for electricitymap, we tried to emulate the behaviour with the thresholds.

Then we figured that people who are using this would probably need the exact numbers and they can decide what is low, moderate and high outside this lib on their own.

**Review by rossf7 — APPROVED** [review]
@ofpiyush Thanks this is great! LGTM

For Electricity Map thanks for the explanation and yes good to have a self hosted option as well. 

> Then we figured that people who are using this would probably need the exact numbers and they can decide what is low, moderate and high outside this lib on their own.

Cool. Again, thanks for the explanation. This makes total sense to me especially at the start. The library can be extended later if people need this.

**Pull request merged**